src/ethereum_clis/ethereum_cli.py

```diff
@@ -78,46 +82,90 @@ def from_binary_path(cls, *, binary_path: Optional[Path], **kwargs) -> Any:
-        binary = shutil.which(binary_path)  # type: ignore
+            logger.debug("Resolved path exists")
+            binary = Path(resolved_path)
+        else:
+            logger.debug(
+                f"Resolved path does not exist: {resolved_path}\nTrying to find it via `which`"
+            )
 
-        if not binary:
-            raise CLINotFoundInPathError(binary=binary)
+            # it might be that the provided binary exists in path
+            filename = os.path.basename(resolved_path)
+            binary = shutil.which(filename)  # type: ignore
+            logger.debug(f"Output of 'which {binary_path}': {binary}")
 
-        binary = Path(binary)  # type: ignore[assignment]
+            if binary is None:
+                logger.error(f"Resolved t8n binary path does not exist: {resolved_path}")
+                raise CLINotFoundInPathError(binary=resolved_path)
+
+            assert binary is not None
+            logger.debug(f"Successfully located the path of the t8n binary: {binary}")
+            binary = Path(binary)
 
         # Group the tools by version flag, so we only have to call the tool once for all the
         # classes that share the same version flag
         for version_flag, subclasses in groupby(
             cls.registered_tools, key=lambda x: x.version_flag
         ):
+            logger.debug(
+                f"Trying this `version` flag to determine if t8n supported: {version_flag}"
+            )
+            # adding more logging reveals we check for `-v` twice..
+
             try:
                 result = subprocess.run(
                     [binary, version_flag], stdout=subprocess.PIPE, stderr=subprocess.PIPE
                 )
+                logger.debug(
+                    f"Subprocess:\n\tstdout: {result.stdout}\n\n\n\tstderr: {result.stderr}\n\n\n"  # type: ignore
+                )
+
                 if result.returncode != 0:
-                    raise Exception(f"Non-zero return code: {result.returncode}")
+                    logger.debug(f"Subprocess returncode is not 0! It is: {result.returncode}")
+                    continue  # don't raise exception, you are supposed to keep trying different version flags  # noqa: E501
 
                 if result.stderr:
-                    raise Exception(f"Tool wrote to stderr: {result.stderr.decode()}")
+                    logger.debug(f"Stderr detected: {result.stderr}")  # type: ignore
+                    continue
 
                 binary_output = ""
                 if result.stdout:
                     binary_output = result.stdout.decode().strip()
-            except Exception:
-                # If the tool doesn't support the version flag,
-                # we'll get an non-zero exit code.
+                    # e.g. 1.31.10+f62cfede9b4abfb5cd62d6f138240668620a2b0d should be treated as 1.31.10  # noqa: E501
+                    if "+" in binary_output:
+                        binary_output = binary_output.split("+")[0]
+
```

Comment:
I think this should be handled by the regex in `detect_binary_pattern`, and with your changes it seems like it already does, so these lines seem redundant.